Making parser service work

TueeNguyen opened the pull request:
<!--
Thanks for sending a pull request!
- if this is your first time, please ensure you have read through our contributor guide: https://github.com/Seneca-CDOT/telescope/blob/master/docs/CONTRIBUTING.md
-->

## Issue This PR Addresses
Fixes #2111 
This is a big change so I'd rather have it looked at early
<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number> <reason>(i.e. This issue was worked on by @user and myself and his PR should close the issue.)
-->

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
- Add changes based on backend and #2706, #2581
- Export Redis constructor instead of Redis instance
- Bump node-fetch to 2.6.7
<!-- Please add a detailed description of what this PR does and why it is needed -->

## Steps to test the PR
1. `pnpm install` in root 
2. `pnpm services:start redis elasticsearch traefik posts`
4. `npm run dev` and go to `localhost:8000` or look at the terminal log
**Notes**: I have not tested this for Youtube feeds, will edit this with test result
## Next steps
Should I open small issues for these steps?
1. Open a proper PR or keep this 
2. Add unit and e2e tests
3. Dockerize parser service


<!-- Please add steps to build the changes in your PR locally so that peers can test your PR
    Example:
    - `npm install`
    - Go to '...'
    - Click on '....'
    - Scroll down to '....'
    - See error

    Please remember, the more clear you are, the faster your PR will be reviewed and merged.
 -->

## Checklist

<!-- Before submitting a PR, address each item -->

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

manekenpix commented on src/api/parser/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# Unsplash photos are downloaded on startup
+photos/*.jpg
+# We have a single photo we use by default until that happens
+!photos/default.jpg
```
You don't need these

JerryHue commented on src/api/parser/README.md:
```diff
@@ -11,10 +11,6 @@ npm install
-npm start
-
-# dev mode with automatic restarts
 npm run dev
```
Let's update this so that it mentions `pnpm`, instead of `npm`.

humphd replied:
When you do, drop the `run` part.

JerryHue reviewed (commented):
I think the dockerization can be done in a separate PR, we want to focus on migrating all the changes from the legacy backend into the parser service, which is already a big task. The testing can be done on a separate PR, as well.

humphd commented on src/api/parser/src/data/feed.js:
```diff
@@ -16,7 +15,11 @@ const {
+
+const { deletePost } = require('../utils/indexer');
+
+const hash = require('./hash');
```
Great catch. Yes, let's pivot to use the one in satellite.

humphd commented:
Can you rebase for the lock file changes?

dbelokon commented on src/api/parser/README.md:
```diff
@@ -10,8 +10,9 @@ pnpm install
 
```
I think the route examples on the README.md do not work anymore.

dbelokon reviewed (commented):
Very nice! I ran it and it filled the redis DB with no issue. I run the front-end with the other services to ensure that some posts would be there (like the YouTube ones), and I had problems searching the posts when I started all services with `pnpm run services:start`. I am not really sure if the problem is related to the parser service.

TueeNguyen commented on src/api/parser/src/index.js:
```diff
@@ -1,20 +1,114 @@
-module.exports = service;
+const { feedQueue } = require('./feed/queue');
+const feedWorker = require('./feed/worker');
+const { logger } = require('./utils/logger');
```
I missed this, thanks!

TueeNguyen commented on src/api/parser/src/index.js:
```diff
@@ -1,20 +1,114 @@
+ * these into Feed Objects to be added to the database and feed queue.
+ */
+const processAllFeeds = async () => {
+  logger.info('how are we doing again');
```
Yes 😂

TueeNguyen commented on src/api/parser/src/utils/storage.js:
```diff
@@ -59,6 +79,30 @@ module.exports = {
 
   getFeeds: () => redis.smembers(feedsKey),
 
+  getInvalidFeeds: async () => {
+    const invalidKeys = await getFeedKeysUsingScanStream(`${feedNamespace}*${invalidSuffix}`);
+    return Promise.all(
+      invalidKeys.map(async (key) => {
+        const reason = await redis.get(key);
+        const id = key.replace(feedNamespace, '').replace(invalidSuffix, '');
+        return {
+          id,
+          reason: reason.replace(/\n/g, ' '),
+        };
+      })
+    );
+  },
+
+  getDelayedFeeds: async () => {
+    const delayedKeys = await getFeedKeysUsingScanStream(`${feedNamespace}*${delayedSuffix}`);
+    return delayedKeys.map((key) => {
+      const id = key.replace(feedNamespace, '').replace(delayedSuffix, '');
+      return {
+        id,
+      };
+    });
+  },
```
It's from the backend, I don't think we should remove either of them

TueeNguyen commented on src/api/parser/src/utils/html/remove-empty-anchor.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = (dom) => {
```
It's also from backend, I don't think it's used anywhere

humphd replied:
Can you wire it up for us in src/api/parser/src/utils/html/index.js?

manekenpix commented on src/api/parser/README.md:
````diff
@@ -10,8 +10,9 @@ pnpm install
+### dev mode with automatic restarts
 
+```
 pnpm start
````
```suggestion
pnpm dev
```

humphd replied:
@TueeNguyen an you make this fix please?

TueeNguyen commented on src/api/parser/src/utils/indexer.js:
```diff
@@ -0,0 +1,173 @@
+
+const { ELASTIC_MAX_RESULTS_PER_PAGE = 5, POSTS_URL = 'http://localhost/v1/posts' } = process.env;
+const { logger } = require('@senecacdot/satellite');
+const { client } = require('../lib/elastic');
```
Should we use `elasticsearch client` from satellite?

manekenpix replied:
Yes, since we're using satellite's `redis` client, I think it'd make sense to do the same with `elasticsearch`.

TueeNguyen commented:
@humphd, I've fixed previous comments. Can you take a look. 
While you're here, I have a question, does `parser` expose any route other services or it just processes feeds?

manekenpix commented:
@TueeNguyen Is this PR supposed to include the `Docker` part? The parser service is still disabled [here](https://github.com/TueeNguyen/telescope/blob/a3105438d21766eb2253aa611219135dbfd379e1/docker/docker-compose.yml#L229).
Also, I don't think you need [this file](https://github.com/TueeNguyen/telescope/blob/feed-parser/src/api/parser/src/server.js) since there's no web server running (and no need to define a port for it).

TueeNguyen commented:
Thanks @manekenpix, I was wondering about the `server.js` file too. The Dockerfile must have slipped in there

TueeNguyen commented:
Got some weird symbols after the time
![image](https://user-images.githubusercontent.com/58532267/154183917-14c1b03c-e9a2-4dd6-9f48-ac584eaad330.png)

manekenpix commented:
@TueeNguyen I think they're supposed to be this

![image](https://user-images.githubusercontent.com/23108901/154189154-0e5ed0cb-9245-4a7a-8105-868fbd779a93.png)

manekenpix commented on config/env.development:
```diff
@@ -226,8 +226,8 @@ LOG_LEVEL=debug
 # Redis Server info, password may be optional (e.g., leave empty if you don't set one)
-#REDIS_URL=redis://127.0.0.1
-#REDIS_PORT=6379
+REDIS_URL=redis://127.0.0.1
+REDIS_PORT=6379
```
I'm not sure if this is going to work considering how `satellite` [creates a redis client](https://github.com/Seneca-CDOT/satellite/blob/743ef81b40fb1d7f07223f78b95a7ba4690fbe0f/src/redis.js#L6).
I don't know if we should change `telescope` to adapt to `satellite` or the other way around.

cc @humphd

TueeNguyen replied:
I don't think we should do this, we talked about it here #2685, I must have uncommented

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "@elastic/elasticsearch": "7.12.0",
+    "@elastic/elasticsearch-mock": "0.3.0",
+    "@senecacdot/satellite": "1.17.0",
+    "@wordpress/wordcount": "2.15.0",
```
I don't think this is needed.  It was for the `/stats` endpoint, which we can remove.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "@bull-board/express": "3.7.0",
+    "@elastic/elasticsearch": "7.12.0",
+    "@elastic/elasticsearch-mock": "0.3.0",
+    "@senecacdot/satellite": "1.17.0",
```
Change this to `"latest"`

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "@elastic/elasticsearch-mock": "0.3.0",
+    "@senecacdot/satellite": "1.17.0",
+    "@wordpress/wordcount": "2.15.0",
+    "babel-jest": "27.3.1",
```
Surely this is a dev dep?

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "babel-jest": "27.3.1",
+    "bull": "3.22.0",
+    "clean-whitespace": "0.1.2",
+    "connect-redis": "6.0.0",
```
I don't think we need this anymore, we don't have sessions with the microservices.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "clean-whitespace": "0.1.2",
+    "connect-redis": "6.0.0",
+    "date-fns": "2.26.0",
+    "docker-compose": "0.23.13",
```
No, we don't need this.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "docker-compose": "0.23.13",
+    "dotenv": "10.0.0",
+    "entities": "3.0.1",
+    "express": "4.17.1",
```
I doubt we need this, should all happen via Satellite.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "dotenv": "10.0.0",
+    "entities": "3.0.1",
+    "express": "4.17.1",
+    "express-healthcheck": "0.1.0",
```
Healthchecks happen in Satellite.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "entities": "3.0.1",
+    "express": "4.17.1",
+    "express-healthcheck": "0.1.0",
+    "express-session": "1.17.2",
```
No sessions anymore.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "express-session": "1.17.2",
+    "express-validator": "6.13.0",
+    "feed": "4.2.2",
+    "helmet": "4.6.0",
```
This happens in Satellite

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "feed": "4.2.2",
+    "helmet": "4.6.0",
+    "highlight.js": "11.3.1",
+    "http-proxy-middleware": "2.0.1",
```
What are we proxying now?

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "highlight.js": "11.3.1",
+    "http-proxy-middleware": "2.0.1",
+    "ioredis": "4.28.0",
+    "ioredis-mock": "5.8.0",
```
None of this redis stuff belongs here, it comes from Satellite.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "jsdom": "18.0.1",
+    "linkify-html": "3.0.5",
+    "linkifyjs": "3.0.5",
+    "node-fetch": "2.6.7",
```
This lives on Satellite now

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "node-fetch": "2.6.7",
+    "normalize-url": "6.0.1",
+    "opml-generator": "1.1.1",
+    "passport": "0.5.0",
```
All this happens in Satellite, no passport stuff.

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,65 @@
+    "opml-generator": "1.1.1",
+    "passport": "0.5.0",
+    "passport-saml": "3.2.0",
+    "pino": "7.3.0",
```
Logging happens in Satellite, remove all the pino stuff.

humphd commented on src/api/parser/src/feed/processor.js:
```diff
@@ -5,22 +5,23 @@
 
 const Parser = require('rss-parser');
-const { logger, fetch } = require('@senecacdot/satellite');
+const fetch = require('node-fetch');
```
Get this from satellite

humphd commented on src/api/parser/src/feed/processor.js:
```diff
@@ -5,22 +5,23 @@
 
 // Check for cached ETag and Last-Modified info on the feed.
-function hasHeaders(feed) {
+const hasHeaders = (feed) => {
```
If the only thing you do in an arrow function is return, you don't need a body: `=> (feed.etag || feed.lastModified)`

humphd commented on src/api/parser/src/index.js:
```diff
@@ -1,20 +1,109 @@
+const { feedQueue } = require('./feed/queue');
+const feedWorker = require('./feed/worker');
+const getWikiFeeds = require('./utils/wiki-feed-parser');
+const shutdown = require('./lib/shutdown');
```
I don't think we need this, Satellite does it.

humphd commented on src/api/parser/src/index.js:
```diff
@@ -1,20 +1,109 @@
+const shutdown = require('./lib/shutdown');
+const Feed = require('./data/feed');
+
+/**
```
Get rid of all these functions, I don't think we need.

humphd commented on src/api/parser/src/lib/elastic.js:
```diff
@@ -1,9 +1,16 @@
+ * and point towards `service.docker.localhost` for querying.
+ */
+
 const { ELASTIC_URL, ELASTIC_PORT } = process.env;
```
We don't want any of this.  Use elastic/mock from Satellite.

cc @rclee91, @JiaHua-Zou

TueeNguyen replied:
I don't think this file is used, I changed the import in `indexer.js` to `const { Elastic } = require('@senecacdot/satellite')`, it should take new changes in Satellite

humphd commented on src/api/parser/src/lib/queue.js:
```diff
@@ -1,5 +1,6 @@
 const Bull = require('bull');
-const { logger, Redis } = require('@senecacdot/satellite');
+const { logger } = require('@senecacdot/satellite');
+const Redis = require('./redis');
```
Use Satellite's

humphd commented on src/api/parser/src/utils/__mocks__/indexer.js:
```diff
@@ -0,0 +1,64 @@
+// Mock storage for our es data
```
@rclee91 and @JiaHua-Zou are the elastic-mock-wizards, let's get their review on this part.

rclee91 replied:
I've been trying to look at it. But since there's no tests I can't really tell if it will work or not. 
As for mocking with ElasticSearch, it doesn't look like we need a __mocks__ folder at all, everything is done with `mock.add`.

humphd commented on src/api/parser/src/utils/storage.js:
```diff
@@ -1,6 +1,10 @@
-const { logger, Redis } = require('@senecacdot/satellite');
+const { logger } = require('@senecacdot/satellite');
+const Redis = require('../lib/redis');
```
Get this from Satellite

TueeNguyen replied:
The Redis exported from `../lib/redis` is imported from Satellite

humphd replied:
What's the value of that extra indirection?  You're not doing anything with it.  Just use Satellite's `Redis` everywhere you need it.

We used to have a module for this because we did setup for you.  Now we don't.

TueeNguyen replied:
I'll remove all the indirections

humphd commented on src/api/parser/src/utils/wiki-feed-parser.js:
```diff
@@ -0,0 +1,103 @@
+const fetch = require('node-fetch');
```
Get this from Satellite

humphd commented:
I retriggered the tests, I think they should pass

TueeNguyen commented:
@humphd I removed unnecessary dependencies, changed a few imports to use Satellite's. 

But removing the `Redis` variables in `env.local` results in this 

![image](https://user-images.githubusercontent.com/58532267/154383331-b347052e-bcbc-4c76-9047-0580cd133ed9.png)

It's looking for host `redis` but the redis server in the container has a different hostname. I wonder how it also works for `posts`

TueeNguyen commented:
@JiaHua-Zou @rclee91, can you guys take a look at the `indexer.js`

humphd commented on src/api/parser/env.local:
```diff
@@ -0,0 +1,36 @@
+
+REDIS_URL=redis://localhost
+REDIS_PORT=6379
+REDIS_PASSWORD=
```
The way it works now, we only need the `REDIS_URL`.

rclee91 commented on src/api/parser/src/utils/indexer.js:
```diff
@@ -0,0 +1,174 @@
+const { setIntervalAsync, clearIntervalAsync } = require('set-interval-async/dynamic');
+
+const { ELASTIC_MAX_RESULTS_PER_PAGE = 5, POSTS_URL = 'http://localhost/v1/posts' } = process.env;
+const { logger } = require('@senecacdot/satellite');
+const { Elastic } = require('@senecacdot/satellite');
```
```suggestion
const { logger, Elastic } = require('@senecacdot/satellite');
```

TueeNguyen replied:
Thanks for this, I went back and look for all similar imports

humphd commented on src/api/parser/package.json:
```diff
@@ -1,28 +1,39 @@
+    "normalize-url": "6.0.1",
+    "rss-parser": "3.12.0",
+    "sanitize-html": "2.5.3",
+    "set-interval-async": "2.0.3"
```
Can we get rid of this and use https://nodejs.org/api/timers.html#timerspromisessetintervaldelay-value-options?

humphd commented on src/api/parser/src/utils/html/sanitize.js:
```diff
@@ -0,0 +1,59 @@
+      a: ['href'],
+    },
+    allowedIframeHostnames: [
+      'www.youtube.com',
```
We have a PR in flight right now that alters this file, let's not forget to update this too or later (@HyperTHD).

TueeNguyen replied:
I'll remember

humphd reviewed (approved):
OK, I'm tempted to take this, and let us move forward on tests.

Great work, and than you for not giving up easily.

TueeNguyen commented:
Thanks everyone, I can write tests for it now :D
